### PR TITLE
`Development`: Improve endpoint analysis

### PR DESCRIPTION
**Pull request by Jan-Thurner**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

### Motivation and Context

The current implementation of the analysis of REST calls and Endpoints was inconsistent in a few places 

### Description

This PR implements the follow-up changes requested in the [previous PR](https://github.com/ls1intum/Artemis/pull/8771). It also enables REST calls that start with a redundant `/` to be assigned to the correct Endpoints.

### Steps for Testing

Prerequisites:
- review the code
- review the action's output. An up-to-date run of the action can be found [here](https://github.com/ls1intum/Artemis/actions/runs/10475329866). If you want to run the action again, it can be triggered manually right [here](https://github.com/ls1intum/Artemis/actions/workflows/analysis-of-endpoint-connections.yml).

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
- [x] Code Review 3
- [x] Code Review 4

## Summary by CodeRabbit

- **New Features**
	- Enhanced clarity in file handling by updating terminology from `fileName` to `filePath` across various components.
	- Improved endpoint analysis accuracy with refined matching logic for REST calls.

- **Bug Fixes**
	- Ensured consistent URI formatting in REST call comparisons.

- **Refactor**
	- Renamed variables and constants for better semantic meaning and maintainability.
	- Streamlined logging mechanism for improved readability.

- **Chores**
	- Optimized workflow execution conditions by refining triggers for GitHub Actions.

### CODE REVIEW

File: supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/EndpointAnalyzer.java [hunk 1]
```diff
@@ -60,10 +60,13 @@ private static void analyzeEndpoints() {
...
                     String endpointURI = endpoint.buildComparableEndpointUri();
-                    List<RestCallInformation> matchingRestCalls = restCallMap.getOrDefault(endpointURI, new ArrayList<>());
+                    List<RestCallInformation> restCallsWithMatchingURI = restCallMap.getOrDefault(endpointURI, new ArrayList<>());
 
                     // Check for wildcard endpoints if no exact match is found
-                    checkForWildcardEndpoints(endpoint, matchingRestCalls, endpointURI, restCallMap);
+                    checkForWildcardEndpoints(endpoint, restCallsWithMatchingURI, endpointURI, restCallMap);
+
+                    List<RestCallInformation> matchingRestCalls = restCallsWithMatchingURI.stream()
+                            .filter(restCall -> restCall.method().toLowerCase().equals(endpoint.getHttpMethod().toLowerCase())).toList();
```
Comment (JohannesStoehr) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/EndpointAnalyzer.java:69]: Are different capitalisations of the URIs actually the same?

Comment (Jan-Thurner) [reply supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/EndpointAnalyzer.java:69]: No, different capitalizations of URIs are different. What is compared here is the HTTP Method. On the client, it is specified with `httpClient.get(...)`, while on the Server it is specified with annotations like `@getMapping(...)`.

File: supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts [hunk 2]
```diff
@@ -38,17 +38,17 @@ class ParsingResult {
...
-    static filesWithRestCalls: { fileName: string, restCalls: RestCall[] }[] = [];
+    static filesWithRestCalls: { filePath: string, restCalls: RestCall[] }[] = [];
     private readonly restCalls: RestCall[] = [];
-    private readonly fileName: string;
+    private readonly filePath: string;
     private readonly ast: TSESTree.Program;
 
     /**
-     * @param fileName - The name of the file being processed.
+     * @param filePath - The name of the file being processed.
```
Comment (pzdr7) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts:47]:
```suggestion
     * @param filePath - The path of the file being processed.
```

File: supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts [hunk 3]
```diff
@@ -108,10 +108,10 @@ export class Postprocessor {
                                         urlEvaluationResult = this.evaluateUrl(node.arguments[0], methodDefinition, node, classBody);
                                     }
 
-                                    const fileName = this.fileName;
+                                    const filePath = this.filePath;
                                     if (urlEvaluationResult.resultType === ParsingResultType.EVALUATE_URL_SUCCESS) {
                                         for (let url of urlEvaluationResult.result) {
-                                            this.restCalls.push({ method, url, line, fileName });
+                                            this.restCalls.push({ method, url, line, filePath: filePath });
```
Comment (pzdr7) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts:114]: If the name of the key is already `filePath`, I think you can just write:

```suggestion
                                            this.restCalls.push({ method, url, line, filePath });
```

File: supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts [hunk 4]
```diff
@@ -108,10 +108,10 @@ export class Postprocessor {
                                         urlEvaluationResult = this.evaluateUrl(node.arguments[0], methodDefinition, node, classBody);
                                     }
 
-                                    const fileName = this.fileName;
+                                    const filePath = this.filePath;
                                     if (urlEvaluationResult.resultType === ParsingResultType.EVALUATE_URL_SUCCESS) {
                                         for (let url of urlEvaluationResult.result) {
```
Comment (pzdr7) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts:113]: Will this work?

```suggestion
                                        for (const url of urlEvaluationResult.result) {
```

File: supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts [hunk 5]
```diff
@@ -61,7 +61,7 @@ export class Postprocessor {
             }
         });
         if (this.restCalls.length > 0) {
-            Postprocessor.filesWithRestCalls.push( {fileName: this.fileName, restCalls: this.restCalls} );
+            Postprocessor.filesWithRestCalls.push( {filePath: this.filePath, restCalls: this.restCalls} );
```
Comment (pzdr7) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/typeScript/Postprocessor.ts:64]:
```suggestion
            Postprocessor.filesWithRestCalls.push({ filePath: this.filePath, restCalls: this.restCalls });
```

File: supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/RestCallInformation.java [hunk 6]
```diff
@@ -12,6 +12,9 @@ public String buildComparableRestCallUri() {
 
         // Remove query parameters
         result = result.split("\\?")[0];
+        if (result.startsWith("/")) {
+            result = result.substring(1);
+        }
```
Comment (pzdr7) [inline supporting_scripts/analysis-of-endpoint-connections/src/main/java/de/tum/cit/endpointanalysis/RestCallInformation.java:17]: What's happening here? Can you add a comment?

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Some small comments on the code

**Review by pzdr7 — APPROVED** [review]
Code

**Review by az108 — APPROVED** [review]
The LGTM after the changes. I also tested it manually in the workflow on the develop branch.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by JohannesWt — APPROVED** [review]
Code 👍

**Review by MarkusPaulsen — APPROVED** [review]
There are still some incorrectly labeled endpoints (e.g. GET https://artemis.ase.in.tum.de/api/public/imprint, which is labeled as unusedEndpoint, although it is used on https://artemis.ase.cit.tum.de/imprint). Nevertheless for this PR, the code looks good and the analysis files are generated.

**Pull request merged**